man/lnxhcrc.5
.\" Macro for inserting an option synopsis string.
.\" .OS <long> [<short>] [args]
.de OS
.  ds args "
.  if !'\\$3'' .as args \fI\\$3\fP
.  if !'\\$4'' .as args \\$4
.  if !'\\$5'' .as args \fI\\$5\fP
.  if !'\\$6'' .as args \\$6
.  if !'\\$7'' .as args \fI\\$7\fP
.  ds result "[
.  if !'\\$2'' .as result \fB\-\\$2\fP|
.  as result \fB\-\-\\$1\fP
.  if !'\\*[args]'' .as result "\ \\*[args]
.  as result "]
\\*[result]
..
.\" Macro for inserting an option description prologue.
.\" .OD <long> [<short>] [args]
.de OD
.  ds args "
.  if !'\\$3'' .as args \fI\\$3\fP
.  if !'\\$4'' .as args \\$4
.  if !'\\$5'' .as args \fI\\$5\fP
.  if !'\\$6'' .as args \\$6
.  if !'\\$7'' .as args \fI\\$7\fP
.  PD 0
.  if !'\\$2'' .IP "\fB\-\\$2\fP \\*[args]" 4
.  if !'\\$1'' .IP "\fB\-\-\\$1\fP \\*[args]" 4
.  PD
..
.\" Macro for inserting a keyword description prologue.
.\" .KY <keyword> <terminal> <non-terminal> <terminal> <non-terminal>
.de KY
\fB\\$1\fP\\$2\fI\\$3\fP\\$4\fI\\$5\fP
..
.\" Macro for inserting code line.
.\" .CL <text>
.de CL
.  ds pfont \fP
.  nh
.  na
.  ft CW
\\$*
.  ft \\*[pfont]
.  ad
.  hy
.  br
..
.\" Macro for inserting a man page reference.
.\" .MP man-page section [suffix]
.de MP
.  nh
.  na
.  BR \\$1 (\\$2)\\$3
.  ad
.  hy
..
.\" Macro for inserting a note.
.\" .NT <text>
.de NT
.  RS 0
.  TP
.  B Note:
\\$*
.  RE
..
.\" Full name of the health checker
.ds lhc "Linux Health Checker
.\" Man page start
.TH lnxhcrc 5 "lnxhc 1.3-1" 2013-12-18 "\*[lhc]"
.
.
.SH NAME
lnxhcrc \- Linux Health Checker configuration file
.
.
.SH DESCRIPTION
The
.B lnxhcrc
file contains configuration information for the \*[lhc] tool (see
.MP lnxhc 1 ).
.PP
The system-wide configuration file is located at
.BR /etc/lnxhcrc .
To change settings for a single user, place a customized copy of this file
at location
.BR ~/.lnxhc/lnxhcrc .
.PP
Lines in the configuration file can either be:
.RS 4
.IP \(bu 2
Empty lines or lines consisting only of white space characters. The \*[lhc]
ignores these lines.
.IP \(bu 2
Comment lines which start with a hash sign ("#"). The \*[lhc] treats these like
empty lines and ignores them.
.IP \(bu 2
Directives in the form
.BR key = value .
You can find the list of valid directives and their description in section
DIRECTIVES.
.RE
.
.SS "Example configuration file"
.RS 4
.CL #
.CL # Example \*[lhc] configuration file
.CL #
.CL
.CL # Database directories
.CL db_path = /usr/lib/lnxhc
.CL db_path = /usr/local/lib/lnxhc
.CL
.CL # Database caching flag
.CL db_caching = 1
.RE
.
.
.SH DIRECTIVES
You can use configuration file directives to change the configuration of the
\*[lhc]. All directives are optional.
.PP
.KY db_path = directory
.RS
Use
.I directory
as database directory.
.PP
Specifies a directory which the \*[lhc] searches for installed checks,
consumers, and profiles. You can repeat this line to specify multiple
directories. The \*[lhc] installs new checks, consumers, and profiles in the
first specified directory.
.PP
The default database directories are
.RS
.B /usr/lib/lnxhc
.RE
and
.RS
.BR /usr/local/lib/lnxhc .
.RE
.PP
.RE
.
.KY db_caching =1|0
.RS
Enable or disable database caching.
.PP
Specify 0 to disable database caching or 1 to enable it. When database
caching is enabled, the \*[lhc] tries to minimize the overhead of running the
lnxhc tool by reading the full contents of each database directory only
when a change has occurred.
.PP
The default value for this setting is 1.
.PP
.RE
.
.
.SH FILES
.TP
/etc/lnxhcrc
System-wide configuration file.
.TP
~/.lnxhc/lnxhcrc
Per-user configuration file. If this file is present, the system-wide
configuration file is not read.
.
.
.SH "SEE ALSO"
.MP lnxhc 1


CHANGES
commit f8177b4cd23fee1c75bd8988c4e1292c481b7c43
Author: Peter Oberparleiter <oberpar@linux.vnet.ibm.com>
Date:   Wed Dec 18 10:39:54 2013 +0100

    lnxhc: Update CONTRIBUTORS file
    
    Signed-off-by: Peter Oberparleiter <peter.oberparleiter@de.ibm.com>
    Signed-off-by: Hendrik Brueckner <brueckner@linux.vnet.ibm.com>

commit 9d6ba8b148b0b27f43839605225fdc68441cc634
Author: Peter Oberparleiter <oberpar@linux.vnet.ibm.com>
Date:   Wed Dec 18 10:19:18 2013 +0100

    checks: Make common helper script consistent
    
    Remove whitespace diff between common helper script.
    
    Signed-off-by: Peter Oberparleiter <peter.oberparleiter@de.ibm.com>
    Signed-off-by: Hendrik Brueckner <brueckner@linux.vnet.ibm.com>

commit 413ed10c142ae8f4cac11f0af4341f492f0fbab2
Author: Peter Oberparleiter <oberpar@linux.vnet.ibm.com>
Date:   Wed Dec 18 10:13:46 2013 +0100

    checks: Remove unused variables
    
    Signed-off-by: Peter Oberparleiter <peter.oberparleiter@de.ibm.com>
    Signed-off-by: Hendrik Brueckner <brueckner@linux.vnet.ibm.com>

commit a00888c77f7238b618cb222d0b0310aa44d901a8
Author: Peter Oberparleiter <oberpar@linux.vnet.ibm.com>
Date:   Tue Dec 17 15:24:08 2013 +0100

    checks: Remove unnecessary imports
    
    Signed-off-by: Peter Oberparleiter <peter.oberparleiter@de.ibm.com>
    Signed-off-by: Hendrik Brueckner <brueckner@linux.vnet.ibm.com>

commit 123e9d055ac9c89c121a163c611400e6ca7fb2a5
Author: Peter Oberparleiter <oberpar@linux.vnet.ibm.com>
Date:   Tue Dec 17 15:17:33 2013 +0100

    checks: Ensure consistent copyright statements
    
    Add missing copyright statements. Also fix copyright year in one
    instance.
    
    Signed-off-by: Peter Oberparleiter <peter.oberparleiter@de.ibm.com>
    Signed-off-by: Hendrik Brueckner <brueckner@linux.vnet.ibm.com>

commit 4d12b3c104df2be96821c3e83dddf66b465a7e24
Author: Peter Oberparleiter <oberpar@linux.vnet.ibm.com>
Date:   Tue Dec 17 14:59:08 2013 +0100

    checks: Remove spaces at end of line
    
    Signed-off-by: Peter Oberparleiter <peter.oberparleiter@de.ibm.com>
    Signed-off-by: Hendrik Brueckner <brueckner@linux.vnet.ibm.com>

commit 93bd843574900fedad835b680ee4a285098bdfa0
Author: Peter Oberparleiter <oberpar@linux.vnet.ibm.com>
Date:   Tue Dec 17 12:23:05 2013 +0100

    checks: Fix check components
    
    Fix inconsistent check components.
    
    Signed-off-by: Peter Oberparleiter <peter.oberparleiter@de.ibm.com>
    Signed-off-by: Hendrik Brueckner <brueckner@linux.vnet.ibm.com>

commit 2776f8e14c00e08fcad682629b9b641b1f844af1
Author: Peter Oberparleiter <peter.oberparleiter@de.ibm.com>
Date:   Tue Dec 17 11:26:04 2013 +0100

    checks: Fix spelling
    
    Fix some spelling.
    
    Signed-off-by: Peter Oberparleiter <peter.oberparleiter@de.ibm.com>
    Signed-off-by: Hendrik Brueckner <brueckner@linux.vnet.ibm.com>

commit ab8221350a01868d03fc1f1d12b543cedd2779d5
Author: Hendrik Brueckner <brueckner@linux.vnet.ibm.com>
Date:   Mon Dec 16 10:30:35 2013 +0100

    lnxhc: Fix misspellings using 'codespell' tool
    
    Signed-off-by: Hendrik Brueckner <brueckner@linux.vnet.ibm.com>
    Signed-off-by: Peter Oberparleiter <peter.oberparleiter@de.ibm.com>

commit a602fb9a5d112ac2acd1cee3c2a0fc7c3e994681
Author: Peter Oberparleiter <peter.oberparleiter@de.ibm.com>
Date:   Fri Dec 13 12:52:44 2013 +0100

    lnxhc: Fix locale handling
    
    System information is collected using the C locale to ensure that
    locale settings do not affect the format of tool output. lnxhc
    uses LANG=C for this purpose, but there are other environment variables
    that influence the language of tool output. Using LC_ALL fixes this
    problem.
    
    Signed-off-by: Peter Oberparleiter <peter.oberparleiter@de.ibm.com>
    Signed-off-by: Hendrik Brueckner <brueckner@linux.vnet.ibm.com>

commit 0663746b4e073b4a83f5930fe061f30d125a91e1
Author: Manik Bajpai <manibajp@in.ibm.com>
Date:   Mon Sep 23 08:36:13 2013 +0200

    zfcp_lun_configured_available: Check configured LUNs for availability
    
    Ensure that all LUNs configured for persistence, which are accessible
    through an FCP adapter, are available
    
    Users configure LUNs when they intend to use them for I/O. If a
    configured LUN is not available, this indicates either a configuration
    error or a loss of connectivity.
    
    Signed-off-by: Manik Bajpai <manibajp@in.ibm.com>
    Signed-off-by: Hendrik Brueckner <brueckner@linux.vnet.ibm.com>
    Signed-off-by: Peter Oberparleiter <peter.oberparleiter@de.ibm.com>

commit b43d4a23dbb041d8bf0a30d927a911b098640b2e
Author: Hendrik Brueckner <brueckner@linux.vnet.ibm.com>
Date:   Mon Nov 11 17:21:50 2013 +0100

    tty_devnodes, tty_usage: load data in a restricted compartment
    
    Use a restricted compartment to load and parse the ls-tty system
    information.
    
    Signed-off-by: Hendrik Brueckner <brueckner@linux.vnet.ibm.com>
    Signed-off-by: Peter Oberparleiter <peter.oberparleiter@de.ibm.com>

commit eb126a5d621e9881cb474de7d15617894823b2ec
Author: Peter Oberparleiter <peter.oberparleiter@de.ibm.com>
Date:   Wed Aug 7 10:55:39 2013 +0200

    lnxhc: prevent data corruption in case of concurrent runs
    
    Data stored in one of lnxhc's databases might become unreadable
    if two processes try to write to them at once. Fix this by using
    locked access to the corresponding files.
    
    Signed-off-by: Peter Oberparleiter <peter.oberparleiter@de.ibm.com>
    Signed-off-by: Hendrik Brueckner <brueckner@linux.vnet.ibm.com>

commit cb6b397ddb7e52dae8f5d7a4905460e2c8d96f5f
Author: Peter Oberparleiter <peter.oberparleiter@de.ibm.com>
Date:   Tue Aug 6 09:10:57 2013 +0200

    lnxhc: updated CONTRIBUTORS file
    
    Signed-off-by: Peter Oberparleiter <peter.oberparleiter@de.ibm.com>
    Signed-off-by: Hendrik Brueckner <brueckner@linux.vnet.ibm.com>

commit b84874bcca30f6ed6d9051fa641193c0e53b61be
Author: Peter Oberparleiter <peter.oberparleiter@de.ibm.com>

lib/LNXHC/LNXHCRC.pm
#
# LNXHC::LNXHCRC.pm
#   Linux Health Checker configuration file handling
#
# Copyright IBM Corp. 2012
#
# Author(s): Peter Oberparleiter <peter.oberparleiter@de.ibm.com>
#
# All rights reserved. This program and the accompanying materials
# are made available under the terms of the Eclipse Public License v1.0
# which accompanies this distribution, and is available at
# http://www.eclipse.org/legal/epl-v10.html
#
# Contributors: See file CONTRIBUTORS which is part of this package
#

package LNXHC::LNXHCRC;

use strict;
use warnings;

use Exporter qw(import);
use File::Spec::Functions qw(catfile rootdir);


#
# Local imports
#
use LNXHC::Consts qw($LNXHCRC_ID_T_DB_CACHING $LNXHCRC_ID_T_DB_PATH);
use LNXHC::Misc qw(info3 unquote_nodie);
use LNXHC::UData qw(udata_get_path);


#
# Default exports
#
our @EXPORT;


#
# On-demand exports
#
our @EXPORT_OK = qw(&lnxhcrc_get);


#
# Constants
#

my $_RC_FILENAME		= "lnxhcrc";

# lnxhcrc item data
# Enumeration of data fields for struct lnxhcrc_item_t
my $_ITEM_T_ID			= 0;
my $_ITEM_T_REPEATS		= 1;
my $_ITEM_T_DEFAULT		= 2;

# Definition of contents of lnxhcrc file
my %_DEF = (
	"db_path" =>
		[
			$LNXHCRC_ID_T_DB_PATH,
			1,
			\@main::default_db_dirs,
		],
	"db_caching" =>
		[
			$LNXHCRC_ID_T_DB_CACHING,
			0,
			1,
		],
);


#
# Global variables
#

my $_lnxhcrc;


#
# Sub-routines
#

#
# _add_defaults()
#
# Set default values for configuration file items which have not yet been
# set.
#
sub _add_defaults()
{
	foreach my $item (values(%_DEF)) {
		my $id = $item->[$_ITEM_T_ID];
		my $default = $item->[$_ITEM_T_DEFAULT];

		# Skip already defined items
		next if (defined($_lnxhcrc->{$id}));

		# Set default value
		$_lnxhcrc->{$id} = $default;
	}
}

#
# _read_lnxhcrc()
#
# Read contents of lnxhc configuration file. Search order:
# - user data directory (either default or specified by --user-data)
# - /etc
#
sub _read_lnxhcrc()
{
	my $filename;
	my $handle;
	my $err;

	$_lnxhcrc = {};

	# Determine location of configuration file
	$filename = udata_get_path($_RC_FILENAME);
	if (!-e $filename) {
		$filename = catfile(rootdir(), "etc", $_RC_FILENAME);
		if (!-e $filename) {
			goto out;
		}
	}
	info3("Using configuration file '$filename'\n");

	# Parse configuration file
	open($handle, "<", $filename) or
		die("Could not read configuration file '$filename': $!\n");
	foreach my $line (<$handle>) {
		my ($key, $value);
		my $item;
		my ($id, $repeats);

		# Skip empty lines
		next if ($line =~ /^\s*$/);
		# Skip comments lines
		next if ($line =~ /^\s*#/);

		# Check format
		if ($line !~ /^\s*(\w+)\s*=(.*)$/) {
			$err = "unrecognized line format";
			goto err;
		}
		# Check keyword
		($key, $value) = ($1, $2);
		if (!exists($_DEF{$key})) {
			$err = "unknown keyword '$key'";
			goto err;
		}
		# Remove quoting from value
		($err, $value ) = unquote_nodie($value);
		if (defined($err)) {
			goto err;
		}
		# Add to hash
		$item = $_DEF{$key};
		($id, $repeats) = @$item;
		if ($repeats) {
			push(@{$_lnxhcrc->{$id}}, $value);
		} else {
			if (exists($_lnxhcrc->{$id})) {
				warn("Configuration keyword '$key' redefined ".
				     "in $filename:$.\n");
			}
			$_lnxhcrc->{$id} = $value;
		}
	}
	close($handle);

out:
	_add_defaults();
	return;

err:
	die("Configuration file error in $filename:$.: $err!\n");
}

#
# lnxhcrc_get(item_id)
#
# Return value for lnxhcrc item with specified ITEM_ID.
#
sub lnxhcrc_get($)
{
	my ($item_id) = @_;

	# Lazy configuration file reading
	_read_lnxhcrc() if (!defined($_lnxhcrc));

	return $_lnxhcrc->{$item_id};
}


#
# Code entry
#

# Indicate successful module initialization
1;
